proposals/ref-lifetimes.md

```diff
@@ -0,0 +1,294 @@
+
+# Ref lifetimes
+
+## Defining safety rules
+
+With the addition of ref fields into the language the safety rules need to be expanded to
+encompass them. Fundamentally, the ref safety rules have only one purpose: to ensure
+that a ref to a location is not referencable after the location's lifetime ends.
+
+All value locations (i.e., things to which you can take a `ref`) have an existing lifetime
+already defined by the C# language. The only interesting augmentations are around ref
+variables.
+
+The basic rule can be defined quite simply: a ref variable's lifetime must never be longer
+than the value it points to. For ref locals, this is trivial. A ref local is always initialized
+to a value expression or another ref expression. For a value expression, its lifetime can be
+assigned the lifetime of the value expression. For a ref expression, the lifetime is the same
+as the ref expression.
+
+Ref parameters and returns are more complicated because they do not have an initializing value
+expression. In fact, different call sites may have completely different and incomparable lifetime
+parameters. If we were to strictly follow the rule that the lifetime for a variable must be no
+longer than the shortest lifetime, there would be no valid lifetime across all call sites.
+
+Instead, we can introduce generic lifetimes, which allow each call site to substitute a given
+lifetime for each lifetime parameter. To represent such a scheme, we must introduce a syntax for
+specifying the lifetime of a ref variable. To differentiate from conventional generics, we will add
+the generic parameters as belonging to the `ref` keyword for ref variables, instead of the type
+syntax. We will also use the `'` token to prefix the names of each parameter. For instance,
+
+```C#
+ref<'a> T M<'a,'b, T>(ref<'a> T p1, ref<'b> int p2) { ... }
```

Comment:
I think the generic lifetime is better suited for being on the `scoped` modifier, not `ref`. The rationale being that when you have a `ref` on a `ref struct` type then there are actually two lifetimes involved: the lifetime of the `ref` and of the value it points to. Even in the C# 11 syntax these can be differentiated by the placement of the `scoped` modifier. 

https://raw.githubusercontent.com/dotnet/csharplang/main/proposals/low-level-struct-improvements.md#rules-scoped

Putting the lifetime on the `ref` modifier means you don't have the maximum flexibility supported by the underlying data.

Reply:
Seems fine to me, but I don't think you need this. My proposal has two additions of where you can put lifetime annotations for values:

1. On the `ref` keyword.
2. At the beginning of a struct's (ref struct or regular) generic parameter list.

This lets you differentiate between the lifetime of the `ref`, and the lifetime of the struct/stuff inside the struct. For example,

```C#
ref struct Span<'a, T> { ... }

// lifetime on Span
Span<'a, int> M<'a>(Span<'a, int> s) { ... }

// lifetime on ref
ref<'a> Span<'global, int> M<'a>(ref<'a> Span<'global, int> s) { ... }